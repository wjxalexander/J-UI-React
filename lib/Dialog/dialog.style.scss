@import "../helper";
.j-ui-dialog {
  &-container {
    position: fixed;
    background: #fff;
    text-align: left;
    max-height: 100%;
    overflow-y: auto;
    box-shadow: rgba(0, 0, 0, 0.22) 0 25.6px 57.6px 0, rgba(0, 0, 0, 0.18) 0 4.8px 14.4px 0;
    width: auto;
    max-width: 25em;
    min-width: 20em;
    padding: 16px 24px 24px 24px;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);

    & .j-ui-dialog-close{
      position: absolute;
      top: 14px;
      right: 14px;
      width: 2em;
      height: 2em;
      display: flex;
      align-items: center;
      justify-content: center;
      &:hover{
        color: $hover_color;
        border-radius: 2px;
        background-color: $hover_background_color;
      }
    }
  }
    &-header{
      font-size: 1.2em;
      font-weight:600;
      padding: 0 24px 24px 0;
    }
    &-main{
      min-height: 6em;

    }
    &-footer{
      position: relative;
      display: flex;
      justify-content: flex-end;
      align-items: center;
      width: 100%;
      min-height: 24px;
      line-height: 24px;
      margin: 16px 0 0;
      span{
        margin-left: 8px;
      }
      //font-size: 0px;
    }

  &-mask {
    position: fixed;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
    background: fade_out(#fff, 0.5);
  }
}