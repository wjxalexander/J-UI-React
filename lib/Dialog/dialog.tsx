import React, {Fragment, ReactElement} from "react"
import "./dialog.style.scss"
import {Icon, Button} from "../index";
import {rootClass} from "../utils/classFactory";

interface dialogProps {
  visible: boolean;
  children?: React.ReactNode
  buttons?: Array<ReactElement>;
  onDismiss: React.MouseEventHandler<HTMLOrSVGElement>;
  maskClosable?: boolean;
}

// button Generation
const buttons = (buttonArray: Array<ReactElement>): ReactElement => {
  return (
    <Fragment>
      {buttonArray.map((item, index) => (<span key={`${index}-${Math.random()}`}>{item}</span>))}
    </Fragment>
  )
};
// HIGH ORDER FUNCTION
const classGenerator = rootClass('j-ui-dialog');


const Dialog: React.FunctionComponent<dialogProps> = (props: dialogProps) => {
  const onDismiss: React.MouseEventHandler<HTMLOrSVGElement> = (event: React.MouseEvent<HTMLOrSVGElement>) => {
    props.onDismiss(event)
  };
  const MaskDismiss: React.MouseEventHandler<HTMLOrSVGElement> = (event: React.MouseEvent<HTMLOrSVGElement>) => {
    if (!props.maskClosable) {
      return
    }
    props.onDismiss(event)
  };
  return (
    props.visible ?
      <Fragment>
        <div className={classGenerator("mask")} onClick={MaskDismiss}/>
        <div className={classGenerator("container")}>
          <Icon iconName={'Cancel'} USEMsFabricIcon
                className={classGenerator("close")}
                onClick={onDismiss}/>
          <header className={classGenerator("header")}>All emails together</header>
          <main className={classGenerator("content")}>{props.children}</main>
          <footer className={classGenerator("footer")}>
            {props.buttons && props.buttons.length > 0 ? (buttons(props.buttons)) :
              (<Fragment>
                <span><Button buttonType="default" title="Save"/></span>
                <span><Button buttonType="custom" title="Cancel"/></span>
              </Fragment>)}
          </footer>
        </div>
      </Fragment> : null

  )
};
Dialog.defaultProps = {
  maskClosable: true
};
export default Dialog