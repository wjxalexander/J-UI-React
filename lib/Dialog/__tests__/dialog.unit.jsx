import React from 'react';
import Dialog,{buttonsGenerate} from '../dialog';
import Enzyme, {mount, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {Icon, Button} from "../../index"

Enzyme.configure({adapter: new Adapter()});

describe('Dialog', () => {
  it('renders correctly', () => {
    let visible = false;
    const fn = jest.fn(() => visible = !visible);
    const fn2 = jest.fn();
    const wrapper = shallow(<Dialog className="Dialog-test" onOk={fn2} onDismiss={fn} visible={visible}>
      <div>123</div>
    </Dialog>);
    expect(wrapper.find('.Dialog-test')).not.toExist();
    fn();
    const newWrapper = shallow(<Dialog className="Dialog-test" onOk={fn2} onDismiss={fn} visible={visible}>
      <div>123</div>
    </Dialog>);
    const defaultButton = newWrapper.find(Button)
    expect(defaultButton.length).toBe(2);
    expect(defaultButton.at(0).prop('title')).toBe('Save');
    expect(defaultButton.at(1).prop('title')).toBe('Cancel');
    defaultButton.at(0).simulate('click');
    expect(fn2).toBeCalled()
    expect(newWrapper.find('.Dialog-test')).toExist();
    expect(newWrapper.find('.j-ui-dialog-mask')).toExist();
    expect(newWrapper.find('.Dialog-test')).toMatchSnapshot();
    expect(newWrapper.prop('buttons')).toBeFalsy();
  });
  it('buttons renders correctly', () => {
    let visible = false;
    const fn = jest.fn(() => visible = !visible);

    const wrapper = mount(<Dialog className="Dialog-test" onDismiss={fn} visible={visible} buttons={[<Button buttonType="default" title="保存"/>,
      <Button onClick={fn} buttonType="custom" title="取消"/>]}>
      <div>123</div>
    </Dialog>);

    const buttonMock = buttonsGenerate([<Button buttonType="default" title="保存"/>,
      <Button onClick={fn} buttonType="custom" title="取消"/>])
    expect(buttonMock).toMatchSnapshot()
    console.log(wrapper.debug())
    expect(wrapper.prop('buttons')).toBeTruthy();
    expect(wrapper.prop('buttons').length).toBe(2);

    //
    // const defaultButton = wrapper.find(Button);
    // expect(defaultButton.at(0).prop('title')).toBe('Save');
    // expect(defaultButton.at(1).prop('title')).toBe('Cancel');
  });
  it('mask dismiss correctly', () => {
      let visible = true;
      const fn = jest.fn(() => visible = !visible);
      const wrapper = shallow(<Dialog className="Dialog-test" onDismiss={fn} visible={visible}/>);
      wrapper.find('.j-ui-dialog-mask').simulate('click');
      expect(fn).toBeCalled();
      expect(visible).toBe(false)
    }
  )
  ;
  it('X dismiss correctly', () => {
    let visible = true;
    const fn = jest.fn(() => visible = !visible);
    const wrapper = shallow(<Dialog className="Dialog-test" onDismiss={fn} visible={visible}/>);
    expect(wrapper.find('.j-ui-dialog-close')).toExist()
    wrapper.find('.j-ui-dialog-close').simulate('click');
    expect(fn).toBeCalled();
    expect(visible).toBe(false);
  });
  it('no X no mask dismiss correctly', () => {
    let visible = true;
    const fn = jest.fn(() => visible = !visible);
    const wrapper = shallow(<Dialog maskClosable={false} showCloseButton={false} className="Dialog-test"
                                    onDismiss={fn} visible={visible}/>);
    expect(wrapper.find('.j-ui-dialog-close')).not.toExist();
    wrapper.find('.j-ui-dialog-mask').simulate('click');
    expect(fn).not.toBeCalled();
  });
  it('default button correctly', () => {
    let visible = true;
    const fn = jest.fn(() => visible = !visible);
    const wrapper = mount(<Dialog className="Dialog-test" onDismiss={fn} visible={visible}/>);
    const buttons = wrapper.find(Button);
    expect(buttons.length).toBe(2);
    wrapper.find('Button[title="Cancel"]').simulate('click');
    expect(fn).toBeCalled();
    expect(visible).toBe(false);
  });
});
