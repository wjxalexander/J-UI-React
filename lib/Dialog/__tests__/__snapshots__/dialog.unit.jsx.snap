// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog buttons renders correctly 1`] = `
<React.Fragment>
  <span>
    <Button
      buttonType="default"
      title="保存"
    />
  </span>
  <span>
    <Button
      buttonType="custom"
      onClick={[MockFunction]}
      title="取消"
    />
  </span>
</React.Fragment>
`;

exports[`Dialog renders correctly 1`] = `
<div
  className="j-ui-dialog-container Dialog-test"
>
  <Icon
    USEMsFabricIcon={true}
    className="j-ui-dialog-close"
    iconName="Cancel"
    onClick={[Function]}
  />
  <header
    className="j-ui-dialog-header"
  >
    提示
  </header>
  <main
    className="j-ui-dialog-content"
  >
    <div>
      123
    </div>
  </main>
  <footer
    className="j-ui-dialog-footer"
  >
    <span>
      <Button
        buttonType="default"
        onClick={[Function]}
        title="Save"
      />
    </span>
    <span>
      <Button
        buttonType="custom"
        onClick={[Function]}
        title="Cancel"
      />
    </span>
  </footer>
</div>
`;
